File: frontend-design/scripts/generate_palette_brief.py
#!/usr/bin/env python3
"""Generate palette guidance from product palettes and color psychology.

Usage:
  python scripts/generate_palette_brief.py fintech crypto
  python scripts/generate_palette_brief.py healthcare --colors blue green
"""

from __future__ import annotations

import argparse
from _catalog import rank_csv, pick, md_bullet


def main() -> int:
    parser = argparse.ArgumentParser(description="Generate color palette brief")
    parser.add_argument("query", nargs="+")
    parser.add_argument("--colors", nargs="*", default=[])
    args = parser.parse_args()
    query = " ".join(args.query)
    palette_rows = rank_csv("product-color-palettes.csv", query, 4)
    psychology_rows = []
    for color in args.colors or query.split():
        psychology_rows.extend(rank_csv("color-psychology.csv", color, 1))
    seen = set()
    psychology_rows = [row for row in psychology_rows if not (pick(row, "color") in seen or seen.add(pick(row, "color")))]
    print(f"# Palette Brief: {query}\n")
    if palette_rows:
        print("## Product Palette Candidates\n")
        for row in palette_rows:
            print(f"### {pick(row, 'Product Type')}")
            for label, key in [
                ("Primary", "Primary"), ("On primary", "On Primary"), ("Secondary", "Secondary"),
                ("On secondary", "On Secondary"), ("Accent", "Accent"), ("On accent", "On Accent"),
                ("Background", "Background"), ("Foreground", "Foreground"),
                ("Muted", "Muted"), ("Border", "Border"), ("Success", "Success"),
                ("Warning", "Warning"), ("Error", "Error"), ("Info", "Info"),
            ]:
                value = pick(row, key)
                if value:
                    print(md_bullet(label, value))
            print("")
    if psychology_rows:
        print("## Color Psychology\n")
        for row in psychology_rows:
            print(f"### {pick(row, 'color')}")
            print(md_bullet("Signals", pick(row, "signals")))
            print(md_bullet("Common fit", pick(row, "common_fit")))
            print(md_bullet("Caution", pick(row, "caution")))
            print(md_bullet("Pairings", pick(row, "pairings")))
            print("")
    print("## Rules\n")
    print("- Define semantic roles before painting components.")
    print("- Check contrast for primary text, muted text, borders, and focus rings.")
    print("- Reserve strong color for action, status, or brand memory.")
    print("- Do not rely on color alone for state or severity.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend-design/scripts/generate_palette_brief.mjs
#!/usr/bin/env node
/** Generate palette guidance from product palettes and color psychology. */

import { pathToFileURL } from "node:url";
import { mdBullet, pick, rankCsv } from "./_catalog.mjs";

const PALETTE_FIELDS = [
  ["Primary", "Primary"], ["On primary", "On Primary"], ["Secondary", "Secondary"], ["On secondary", "On Secondary"],
  ["Accent", "Accent"], ["On accent", "On Accent"], ["Background", "Background"], ["Foreground", "Foreground"],
  ["Muted", "Muted"], ["Border", "Border"], ["Success", "Success"], ["Warning", "Warning"], ["Error", "Error"], ["Info", "Info"],
];

function parseArgs(args) {
  const query = [];
  const colors = [];
  let readingColors = false;
  for (const argument of args) {
    if (argument === "--colors") readingColors = true;
    else if (readingColors) colors.push(argument);
    else query.push(argument);
  }
  if (query.length === 0) throw new Error("Usage: generate_palette_brief.mjs <query> [--colors <color...>]");
  return { query: query.join(" "), colors };
}

export function main(args = process.argv.slice(2)) {
  let options;
  try {
    options = parseArgs(args);
  } catch (error) {
    console.error(error.message);
    return 2;
  }
  const palettes = rankCsv("product-color-palettes.csv", options.query, 4);
  const psychology = [];
  for (const color of options.colors.length > 0 ? options.colors : options.query.split(" ")) psychology.push(...rankCsv("color-psychology.csv", color, 1));
  const seen = new Set();
  const uniquePsychology = psychology.filter((row) => {
    const color = pick(row, "color");
    if (seen.has(color)) return false;
    seen.add(color);
    return true;
  });
  const lines = [`# Palette Brief: ${options.query}`, ""];
  if (palettes.length > 0) {
    lines.push("## Product Palette Candidates", "");
    for (const row of palettes) {
      lines.push(`### ${pick(row, "Product Type")}`);
      for (const [label, key] of PALETTE_FIELDS) {
        const value = pick(row, key);
        if (value) lines.push(mdBullet(label, value));
      }
      lines.push("");
    }
  }
  if (uniquePsychology.length > 0) {
    lines.push("## Color Psychology", "");
    for (const row of uniquePsychology) {
      lines.push(
        `### ${pick(row, "color")}`,
        mdBullet("Signals", pick(row, "signals")),
        mdBullet("Common fit", pick(row, "common_fit")),
        mdBullet("Caution", pick(row, "caution")),
        mdBullet("Pairings", pick(row, "pairings")),
        "",
      );
    }
  }
  lines.push(
    "## Rules", "",
    "- Define semantic roles before painting components.",
    "- Check contrast for primary text, muted text, borders, and focus rings.",
    "- Reserve strong color for action, status, or brand memory.",
    "- Do not rely on color alone for state or severity.",
  );
  process.stdout.write(`${lines.join("\n")}\n`);
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) process.exitCode = main();
